internal/services/resource/resource_group_resource.go

```diff
@@ -130,32 +131,73 @@ func resourceResourceGroupDelete(d *pluginsdk.ResourceData, meta interface{}) er
 	if meta.(*clients.Client).Features.ResourceGroup.PreventDeletionIfContainsResources {
 		resourceClient := meta.(*clients.Client).Resource.ResourcesClient
-		// Resource groups sometimes hold on to resource information after the resources have been deleted. We'll retry this check to account for that eventual consistency.
-		err = pluginsdk.Retry(10*time.Minute, func() *pluginsdk.RetryError {
-			results, err := resourceClient.ListByResourceGroupComplete(ctx, id.ResourceGroup, "", "provisioningState", utils.Int32(500))
-			if err != nil {
-				return pluginsdk.NonRetryableError(fmt.Errorf("listing resources in %s: %v", *id, err))
+		// List API returns cached resource metadata, which sometimes incorrectly returns resources that are already removed.
+		// Check if the resources has been deleted and only exist in cache of the resource group, or there actually exist undeleted resources.
+		results, err := resourceClient.ListByResourceGroupComplete(ctx, id.ResourceGroup, "", "provisioningState", utils.Int32(500))
+		if err != nil {
+			return fmt.Errorf("listing resources in %s: %v", *id, err)
+		}
+
+		existResourceIds := make([]string, 0)             // GET resource returned http 200, resource does exist.
+		nonExistResourceIds := make([]string, 0)          // GET resource returned http 404, resource does not exist.
+		unknownApiVersionResourceIds := make([]string, 0) // resources unable to get API version or GET returned error except 200/404
+		providersClient := meta.(*clients.Client).Resource.ResourceProvidersClient
+
+		for results.NotDone() {
+			val := results.Value()
+			if val.ID != nil {
+				if resourceApiVersion, err := getResourceTypeApiVersion(ctx, providersClient, val); err != nil {
+					log.Printf("[WARN] Get Resource Type Api version failed: %+v.", err)
+					unknownApiVersionResourceIds = append(unknownApiVersionResourceIds, *val.ID)
+				} else if exist, err := resourceExistById(ctx, resourceClient, *val.ID, resourceApiVersion); err != nil {
+					log.Printf("[WARN] Get Resource (%s) failed: %+v.", *val.ID, err)
+					unknownApiVersionResourceIds = append(unknownApiVersionResourceIds, *val.ID)
+				} else {
+					if exist {
+						existResourceIds = append(existResourceIds, *val.ID)
+					} else {
+						nonExistResourceIds = append(nonExistResourceIds, *val.ID)
+					}
+				}
 			}
-			nestedResourceIds := make([]string, 0)
-			for results.NotDone() {
-				val := results.Value()
-				if val.ID != nil {
-					nestedResourceIds = append(nestedResourceIds, *val.ID)
+
+			if err := results.NextWithContext(ctx); err != nil {
+				return fmt.Errorf("retrieving next page of nested items for %s: %+v", id, err)
+			}
+		}
+
+		if len(existResourceIds) > 0 {
+			return resourceGroupContainsItemsError(id.ResourceGroup, append(existResourceIds, unknownApiVersionResourceIds...))
+		}
+
+		if len(unknownApiVersionResourceIds) > 0 {
+			err = pluginsdk.Retry(10*time.Minute, func() *pluginsdk.RetryError {
```

Comment:
this should be using a WaitCheckFunc which handles errors etc - rather than a fixed 10m timeout, that way we can use the Delete timeout to poll as short/long as required

Reply:
Does "WaitCheckFunc" means StateChangeConf ?

Reply:
yes i believe that is what he means